File: backend/services/stigmer-server/pkg/domain/mcpserver/oauth/token.go
package oauth

import (
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

// TokenResponse holds the response from an OAuth token endpoint.
type TokenResponse struct {
	AccessToken  string `json:"access_token"`
	TokenType    string `json:"token_type"`
	ExpiresIn    int64  `json:"expires_in"`
	RefreshToken string `json:"refresh_token,omitempty"`
	Scope        string `json:"scope,omitempty"`

	// Slack V2 nests user tokens under authed_user when only user scopes
	// are requested. Standard OAuth providers leave this nil.
	AuthedUser *authedUser `json:"authed_user,omitempty"`
}

// resolveFromAuthedUser promotes authed_user fields to the top level
// when present.
//
// Slack's V2 OAuth returns a bot token at the top-level access_token
// and the actual user token inside authed_user.access_token. The
// authed_user block is only present when user-level scopes were
// granted, so its presence is a reliable signal that the user token
// should be preferred over the bot token. Standard OAuth providers
// never include authed_user, so the top-level token is used unchanged
// for those.
func (r *TokenResponse) resolveFromAuthedUser() {
	if r.AuthedUser == nil {
		return
	}
	if r.AuthedUser.AccessToken != "" {
		r.AccessToken = r.AuthedUser.AccessToken
	}
	if r.AuthedUser.TokenType != "" {
		r.TokenType = r.AuthedUser.TokenType
	}
	if r.AuthedUser.Scope != "" {
		r.Scope = r.AuthedUser.Scope
	}
}

type authedUser struct {
	AccessToken string `json:"access_token"`
	TokenType   string `json:"token_type"`
	Scope       string `json:"scope"`
}

var tokenHTTPClient = &http.Client{Timeout: 15 * time.Second}

// Token-endpoint client authentication methods, in the RFC 8414
// token_endpoint_auth_methods_supported vocabulary. This package speaks
// the raw RFC strings (like dcr.go's "none"); callers holding the
// OAuthAppSpec proto enum map it at their seam.
//
// An empty or unrecognized method falls back to Basic — the RFC 6749
// §2.3.1 baseline and this package's historical behavior.
const (
	TokenAuthMethodBasic = "client_secret_basic"
	TokenAuthMethodPost  = "client_secret_post"
)

// ExchangeCode exchanges an authorization code for tokens at the given
// token endpoint using the authorization_code grant with PKCE.
//
// For public clients (DCR), clientSecret should be empty.
// For confidential clients (vendor OAuth), clientSecret is from the
// OAuthApp and tokenAuthMethod selects how it is presented (see
// doTokenRequest).
func ExchangeCode(
	ctx context.Context,
	tokenEndpoint string,
	code string,
	redirectURI string,
	codeVerifier string,
	clientID string,
	clientSecret string,
	tokenAuthMethod string,
) (*TokenResponse, error) {
	params := url.Values{
		"grant_type":    {"authorization_code"},
		"code":          {code},
		"redirect_uri":  {redirectURI},
		"code_verifier": {codeVerifier},
		"client_id":     {clientID},
	}

	return doTokenRequest(ctx, tokenEndpoint, params, clientID, clientSecret, tokenAuthMethod)
}

// RefreshToken exchanges a refresh token for a new access token.
//
// For public clients (DCR), clientSecret should be empty.
// For confidential clients (vendor OAuth), clientSecret is from the
// OAuthApp and tokenAuthMethod selects how it is presented (see
// doTokenRequest).
//
// The response may include a new refresh token (token rotation). Callers
// should always check TokenResponse.RefreshToken and update storage if
// a new one is issued.
func RefreshToken(
	ctx context.Context,
	tokenEndpoint string,
	refreshToken string,
	clientID string,
	clientSecret string,
	tokenAuthMethod string,
) (*TokenResponse, error) {
	params := url.Values{
		"grant_type":    {"refresh_token"},
		"refresh_token": {refreshToken},
		"client_id":     {clientID},
	}

	return doTokenRequest(ctx, tokenEndpoint, params, clientID, clientSecret, tokenAuthMethod)
}

func doTokenRequest(
	ctx context.Context,
	tokenEndpoint string,
	params url.Values,
	clientID string,
	clientSecret string,
	tokenAuthMethod string,
) (*TokenResponse, error) {
	// Exactly one credential channel per request — RFC 6749 §2.3 forbids
	// presenting the secret through more than one method, and some servers
	// reject requests that do. Post mode rides the form body; anything
	// else (including empty) is the Basic-header baseline.
	usePostSecret := clientSecret != "" && tokenAuthMethod == TokenAuthMethodPost
	if usePostSecret {
		params.Set("client_secret", clientSecret)
	}

	req, err := http.NewRequestWithContext(
		ctx, http.MethodPost, tokenEndpoint,
		strings.NewReader(params.Encode()),
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create token request: %w", err)
	}
	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")

	if clientSecret != "" && !usePostSecret {
		req.SetBasicAuth(clientID, clientSecret)
	}

	resp, err := tokenHTTPClient.Do(req)
	if err != nil {
		return nil, fmt.Errorf("token request to %s failed: %w", tokenEndpoint, err)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf("failed to read token response: %w", err)
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf(
			"token endpoint %s returned HTTP %d: %s",
			tokenEndpoint, resp.StatusCode, truncateBody(body),
		)
	}

	var tokenResp TokenResponse
	if err := json.Unmarshal(body, &tokenResp); err != nil {
		return nil, fmt.Errorf("failed to parse token response: %w", err)
	}

	tokenResp.resolveFromAuthedUser()

	if tokenResp.AccessToken == "" {
		return nil, fmt.Errorf("token response from %s is missing access_token", tokenEndpoint)
	}

	return &tokenResp, nil
}

File: backend/services/stigmer-server/src/domain/mcpserver/oauth/token.ts
/**
 * OAuth token-endpoint client — ports pkg/domain/mcpserver/oauth/token.go
 * whole: the pre-flight refresh (#17) and the authorization-code exchange
 * (#19). Proven by mcpserver-oauth.conformance.test.ts and the Class B
 * mcpserver-connect suite.
 */
import { truncateBody } from "./truncate-body.js";

/** Token-endpoint response (Go TokenResponse). */
export interface TokenResponse {
  accessToken: string;
  tokenType: string;
  expiresIn: number;
  refreshToken: string;
  scope: string;
}

/**
 * Token-endpoint client authentication methods, in the RFC 8414
 * token_endpoint_auth_methods_supported vocabulary. An empty or
 * unrecognized method falls back to Basic — the RFC 6749 §2.3.1 baseline.
 */
export const TOKEN_AUTH_METHOD_BASIC = "client_secret_basic";
export const TOKEN_AUTH_METHOD_POST = "client_secret_post";

/** Go's tokenHTTPClient 15s timeout — a named constant per guidelines. */
export const TOKEN_REQUEST_TIMEOUT_MS = 15_000;

/**
 * Exchanges an authorization code for tokens using the
 * authorization_code grant with PKCE (Go ExchangeCode). For public
 * clients (DCR), clientSecret is empty; for confidential clients (vendor
 * OAuth) tokenAuthMethod selects how the secret is presented.
 */
export async function exchangeCode(
  tokenEndpoint: string,
  code: string,
  redirectUri: string,
  codeVerifier: string,
  clientId: string,
  clientSecret: string,
  tokenAuthMethod: string,
  fetchImpl: typeof fetch = fetch,
): Promise<TokenResponse> {
  const params = new URLSearchParams({
    grant_type: "authorization_code",
    code,
    redirect_uri: redirectUri,
    code_verifier: codeVerifier,
    client_id: clientId,
  });
  return doTokenRequest(
    tokenEndpoint,
    params,
    clientId,
    clientSecret,
    tokenAuthMethod,
    fetchImpl,
  );
}

/**
 * Exchanges a refresh token for a new access token (refresh_token
 * grant). For public clients (DCR), clientSecret is empty; for
 * confidential clients (vendor OAuth) tokenAuthMethod selects how the
 * secret is presented.
 */
export async function refreshToken(
  tokenEndpoint: string,
  currentRefreshToken: string,
  clientId: string,
  clientSecret: string,
  tokenAuthMethod: string,
  fetchImpl: typeof fetch = fetch,
): Promise<TokenResponse> {
  const params = new URLSearchParams({
    grant_type: "refresh_token",
    refresh_token: currentRefreshToken,
    client_id: clientId,
  });
  return doTokenRequest(
    tokenEndpoint,
    params,
    clientId,
    clientSecret,
    tokenAuthMethod,
    fetchImpl,
  );
}

async function doTokenRequest(
  tokenEndpoint: string,
  params: URLSearchParams,
  clientId: string,
  clientSecret: string,
  tokenAuthMethod: string,
  fetchImpl: typeof fetch,
): Promise<TokenResponse> {
  // Exactly one credential channel per request — RFC 6749 §2.3 forbids
  // presenting the secret through more than one method. Post mode rides
  // the form body; anything else (including empty) is the Basic-header
  // baseline.
  const usePostSecret =
    clientSecret !== "" && tokenAuthMethod === TOKEN_AUTH_METHOD_POST;
  if (usePostSecret) {
    params.set("client_secret", clientSecret);
  }

  const headers: Record<string, string> = {
    "Content-Type": "application/x-www-form-urlencoded",
    Accept: "application/json",
  };
  if (clientSecret !== "" && !usePostSecret) {
    headers.Authorization = `Basic ${Buffer.from(`${clientId}:${clientSecret}`).toString("base64")}`;
  }

  let response: Response;
  try {
    response = await fetchImpl(tokenEndpoint, {
      method: "POST",
      headers,
      body: params.toString(),
      signal: AbortSignal.timeout(TOKEN_REQUEST_TIMEOUT_MS),
    });
  } catch (error) {
    throw new Error(
      `token request to ${tokenEndpoint} failed: ${error instanceof Error ? error.message : String(error)}`,
    );
  }

  const body = await response.text();
  if (response.status !== 200) {
    throw new Error(
      `token endpoint ${tokenEndpoint} returned HTTP ${response.status}: ${truncateBody(body)}`,
    );
  }

  let raw: {
    access_token?: string;
    token_type?: string;
    expires_in?: number;
    refresh_token?: string;
    scope?: string;
    authed_user?: {
      access_token?: string;
      token_type?: string;
      scope?: string;
    };
  };
  try {
    raw = JSON.parse(body) as typeof raw;
  } catch (error) {
    throw new Error(
      `failed to parse token response: ${error instanceof Error ? error.message : String(error)}`,
    );
  }

  const tokenResponse: TokenResponse = {
    accessToken: raw.access_token ?? "",
    tokenType: raw.token_type ?? "",
    expiresIn: raw.expires_in ?? 0,
    refreshToken: raw.refresh_token ?? "",
    scope: raw.scope ?? "",
  };

  // Slack V2 nests user tokens under authed_user when only user scopes
  // are requested; its presence is the reliable signal to prefer the user
  // token over the top-level bot token. Standard providers omit it.
  if (raw.authed_user !== undefined) {
    if ((raw.authed_user.access_token ?? "") !== "") {
      tokenResponse.accessToken = raw.authed_user.access_token as string;
    }
    if ((raw.authed_user.token_type ?? "") !== "") {
      tokenResponse.tokenType = raw.authed_user.token_type as string;
    }
    if ((raw.authed_user.scope ?? "") !== "") {
      tokenResponse.scope = raw.authed_user.scope as string;
    }
  }

  if (tokenResponse.accessToken === "") {
    throw new Error(
      `token response from ${tokenEndpoint} is missing access_token`,
    );
  }
  return tokenResponse;
}
